Implementing Sequences

leosartaj opened the pull request:
A **[Sequence](https://github.com/sympy/sympy/wiki/GSoC-2015-Application-Sartaj-Singh:-Improving-the-series-package-and-limits-in-SymPy#sequences)** is finite or infinite lazily evaluated list. Coefficients will only be evaluated when required.
### TODO
- [x] SeqBase Class (abstract class)
- [x] EmptySequence
- [x] SeqExpr class (SeqPer, SeqFunc, SeqFormula inherit from it)
- [x] SeqPer class
- [x] SeqFormula class
- [x] SeqFunc class
- [x] Wrapper function for the above sequences
- [x] Implement SeqExprOp class
- [x] Term-wise Addition/Subtraction
- [x] Term-wise Multiplication
- [x] printing

Ping @jcrist @flacjacket @certik

leosartaj committed:
first version of SeqBase class

This commit introduces sequences module in sympy.series
A sequence is a finite or infinite lazily evaluated list.
SeqBase will be the base class for all the other sequences.

leosartaj committed:
Implemented EmptySequence class

Represents an empty sequence. The empty sequence is available as a
singleton as S.EmptySequence.

leosartaj committed:
Implemented SeqExpr class

Sequence expression class
Various sequences (SeqPer, SeqFormula, SeqFunc...) should inherit from
this class

```
>>> from sympy.series.sequences import SeqExpr
>>> s = SeqExpr((1, 2, 3), (0, 10))
>>> s.gen
(1, 2, 3)
>>> s.interval
[0, 10]
>>> s.length
11
```

leosartaj committed:
Sequences are iterable, step introduced

Futher changes include:
* Helper functions added to SeqBase
* Docs improved
* More tests

leosartaj committed:
Implemented SeqPer class

Represents a periodical sequence
The elements are repeated after a given period.

```
>>> from sympy import SeqPer, oo
>>> s = SeqPer((1, 2, 3), (0, 5))
>>> s.periodical
(1, 2, 3)
>>> s.period
3
```

For value at a particular point

```
>>> s.coeff(3)
1
```

supports slicing

```
>>> s[:]
[1, 2, 3, 1, 2, 3]
```

iterable

```
>>> list(s)
[1, 2, 3, 1, 2, 3]
```

changing step size

```
>>> SeqPer((1, 2, 3), (0, 5, 2))[:]
[1, 3, 2]
```

sequence starts from negative infinity

```
>>> s = SeqPer((1, 2, 3), (-oo, 0))
>>> s[0:6]
[1, 2, 3, 1, 2, 3]
```

leosartaj committed:
Implemented SeqFormula Class

Represents sequence based on a formula
Elements are generated using a formula

```
>>> from sympy import SeqFormula, oo, Symbol
>>> n = Symbol('n')
>>> s = SeqFormula((n**2, n), (0, 5))
>>> s.formula
n**2
```

For value at a particular point

```
>>> s.coeff(3)
9
```

supports slicing

```
>>> s[:]
[0, 1, 4, 9, 16, 25]
```

iterable

```
>>> list(s)
[0, 1, 4, 9, 16, 25]
```

changing step size

```
>>> SeqFormula((n**2, n), (0, 5, 2))[:]
[0, 4, 16]
```

sequence starts from negative infinity

```
>>> s = SeqFormula((n**2, n), (-oo, 0))
>>> s[0:6]
[0, 1, 4, 9, 16, 25]
```

leosartaj committed:
Implemented SeqFunc class

Represents sequence based on a function
Elements are generated by calling a function.
Only single argument functions are allowed.

```
>>> from sympy import SeqFunc, oo, Lambda, Symbol
>>> n = Symbol('n')
>>> s = SeqFunc(Lambda(n, n**2), (0, 5))
>>> s.function
Lambda(n, n**2)
```

For value at a particular point

```
>>> s.coeff(3)
9
```

supports slicing

```
>>> s[:]
[0, 1, 4, 9, 16, 25]
```

iterable

```
>>> list(s)
[0, 1, 4, 9, 16, 25]
```

changing step size

```
>>> SeqFunc(Lambda(n, n**2), (0, 5, 2))[:]
[0, 4, 16]
```

sequence starts from negative infinity

```
>>> SeqFunc(Lambda(n, n**2), (-oo, 0))[0:6]
[0, 1, 4, 9, 16, 25]
```

toolforger commented:
Some quick bikeshedding: "Sequence" is a good name for what you're doing, but the term is already too overloaded with other, inconsistent meanings. I'd stick with "LazyList" or something like that.
I'm not going to insist on that though.

My knee-jerk reaction was "is it using the standard mechanisms in Python, like Iterators?"
Then I saw it's using generators.
I'd like somebody with a higher Python Iterators belt than mine check whether this is typical Python idioms. Not because that's better or anything, but because that would make make it easier for future maintainers if it's following typical Python patterns.

toolforger commented:
> It's part of my GSoC project, some more [details](https://github.com/sympy/sympy/wiki/GSoC-2015-Application-Sartaj-Singh:-Improving-the-series-package-and-limits-in-SymPy).

Ah I see. Laziness is a pretty big thing, as it can have 
counterintuitive effects.

For example, it allows infinite lists, which sounds neat until you 
realize that comparing infinite lists for equality is undecidable.

 From SymPy's perspective, undecidability isn't a big issue, we need to 
deal with it anyway. Usually we go the easy route: punt, abort whatever 
we're doing, and leave the expression unchanged, try a different 
strategy, or do whatever else is the route of least damage.
If your code already does equality testing on infinite lists, you'll 
need to guard against infinite iteration/recursion when iterating over 
the elements of a series.

jcrist commented on sympy/series/__init__.py:
```diff
@@ -5,7 +5,6 @@
 
 O = Order
-
-__all__ = ['gruntz', 'limit', 'series', 'O', 'Order', 'Limit', "residue"]
```
Why did you delete this, instead of adding your stuff to it? This controls what's actually imported from the submodules.

leosartaj replied:
Oh, a mistake on my part. Will fix this in my next commit.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+        return self._gen
+
+    @property
+    def _gen(self):
```
Why the two layers of indirection? Why not just do:

```
@property
def gen(self):
    """Returns the generator for the sequence"""
    raise NotImplementedError(...)
```

jcrist replied:
This goes for all of the below.

leosartaj replied:
I tried to follow the rule, never override a public method. Only private methods should be implemented. I find it cleaner. There is no need to rewrite docstrings. Though I agree, ``gen`` is a little vague. Maybe it's best to drop it altogether and use ``self.args[0]`` everywhere. It's just an other name of ``self.args[0]`` afterall.

flacjacket replied:
I do see the reason you're implementing all the methods as public/private, tho I'm not sure overriding public methods from an abstract class is a bad thing. Also, in some cases it may be good to write new docstrings/doctests because the implementation of these properties varies slightly from one type of sequence to another. Since it is plain that this is a base class that shouldn't be used and anything you actually need will be implemented in subclasses, my vote would be to drop the `_foo` properties and move them to `foo`.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+        return self._length
+
+    @property
+    def _length(self):
```
length is usually just a magic method (i.e. `__len__`). This way `len(sequence)` works.

jcrist replied:
Actually, turns out `len` requires `__len__` to output something that can be an integer. So symbolic lengths won't work (this seems silly to me). So perhaps your method before was fine? @flacjacket : thoughts?

leosartaj replied:
Only because of this reason, I defined a length method.

flacjacket replied:
Maybe in finite, non-symbolic sequences sequences, `__len__` can return reasonable values, it turns on nice features under the hood, but here length is probably ok.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+    return (Interval(start, stop), sympify(step))
+
+
+class SeqExpr(SeqBase):
```
I'm still not sure about your class hierarchy. If you plan on implementing `SeqAdd`, `SeqMul`, or other such classes, then maybe a base class like this makes sense. Otherwise, `SeqBase` can probably just hold everything.

leosartaj replied:
I plan to implement ``SeqAdd and ``SeqMul``. ``Interval``, ``step`` sizes will need to be handled. So, In my opinion, it's good to implement them.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+
+
+class SeqFunc(SeqExpr):
+    """Represents sequence based on a function
```
This doesn't actually work with a function, as I'd expect it to. `SeqFunc(lambda n: n + 1, [0, 1])` raises an error due to `sympify`. If this shouldn't support callables, just `Lambda`, then the docs should be changed. Personally, I can see the desire for having a callable, so any function should probably be allowed.

jcrist replied:
Since normal callables can't be sympified, perhaps any callable isn't the best idea, and only `Lambda` function should be used. Some more thought needs to be put into this.

leosartaj replied:
Yes, only ``Lambda`` functions must be used. I will update the docs. Also, I am having second thoughts on the usefulness of this type of sequence. After implementing it, I don't see it much different from ``SeqFormula``. Or is it okay, to give some versatility?

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+        return iter([])
+
+
+def _parse_interval(interval):
```
Why don't you just use `Range` instead of `Interval`:

```
>>> Range(0, 10)
Range(0, 10, 1)
>>> Range(0, 10, 2)
Range(0, 10, 2)
```

leosartaj replied:
That can work. I was confused, on what would symbolic interval mean. If we were to go with symbolic limits. Then ``Range`` will fail. I have not thought much on that part. Maybe it's best to keep intervals with integer values. In that case ``Range`` is a good option. Opinions? @jcrist @flacjacket

leosartaj replied:
Also, ``Range(0, 5)`` does not include 5. But I would like ``SeqPer((1, 2, 3), (0, 5))`` to include point 5(going by the convention used in ``Sum``). To overcome this maybe ``Range(0, 6)`` can be made internally. Then again, ``Range`` is very restrictive.

flacjacket replied:
Yeah, I don't think you can use `Range` if you need to be able to use symbolic inputs.

flacjacket replied:
As a different note, if you give something like `(0, 5)` as the indices you want, I'd assume you only get indices 0 to 4, just by convention with other Python tools.

leosartaj replied:
Well, I took ``Sum`` as a reference. For something like ``Sum(x**2, (x, 0, 5))`` 5 is included, so going by that convention I decided to include the end point as well.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+
+    is_iterable = True
+
+    def __new__(cls, gen, interval=(0, S.Infinity, 1)):
```
You'll probably need to implement custom `__new__` (or have a `_validate` or something for each one). Currently gen can be anything I put in, and the args aren't uniform. See examples for each class below.

leosartaj replied:
``_validate`` method is a good idea.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+
+    sympy.series.sequences.SeqFormula
+    sympy.series.sequences.SeqFunc
+    """
```
Things that currently work for this, that shouldn't:

```
>>> SeqPer((0, 1, 2, 3), [0, 2])      # More values than interval is long
>>> SeqPer(Range(0, 10, 1), [0, 2])   # Periodic can be whatever you put in - no validation
>>> SeqPer([1, 2, 3], [0, 5, 2])      # Lists aren't converted to tuples, not hashable
```

jcrist replied:
Actually, perhaps the first one shouldn't be forbidden. Thoughts on this?

leosartaj replied:
I say we should allow the first one. I agree all values will not be reached. But then there is need to be that strict. Also, this leniency maybe useful when performing operations.

```
>>> SeqPer((0, 1, 2, 3), (0, 3)) + SeqPer((0, 1, 2, 3), (0, 2))
SeqPer((0, 1, 2, 3), (0, 2))
```
I would work on the validation part.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+
+    @property
+    def _dummy(self):
+        return (self.gen[1],)
```
Why is this in a tuple? Also, perhaps a better name than `dummy` could be used. `variable`?

leosartaj replied:
``variable`` will be better. ``Tuple`` is for the only reason that I can loop over it's content. This was done to be more flexible. In case there are more than one dummy variables.

jcrist commented:
Looks like a good start! I think you need to think about the design of these classes a little bit more. Specific things:
1. What should be allowed in `args` for each type of `Sequence`. How should these `args` be ordered?
2. What inputs should be allowed for each type of `Sequence`?
3. What validation needs to be done for each type of `Sequence` to ensure no invalid state is possible?
4. How can we share as much functionality between these classes as possible?

leosartaj commented:
1. I have followed the design, that for every sequence, there will be first a generator(generates the sequence) and second the interval. Do you see any problems with this design? Well, I have tried to stick with all args being ``Sympy`` objects.
2. Each sequence allows a generator and an interval only. Generator can be different for different sequences. Though, interval has a consistent structure throughout. 
3. I will work more  on the validation part.
4. For this purpose, I decided to create ``SeqExpr``, so that common stuff can be abstracted out.

leosartaj committed:
Implemented sequence wrapper function

Returns appropriate sequence object.

```
>>> from sympy import sequence, SeqPer, SeqFunc, SeqFormula, Lambda
>>> from sympy.abc import n

>>> sequence(formula=(n**2, n), interval=(0, 5))
SeqFormula((n**2, n), ([0, 5], 1))

>>> sequence(periodical=(1, 2, 3), interval=(0, 5))
SeqPer((1, 2, 3), ([0, 5], 1))

>>> sequence(func=Lambda(n, n**2), interval=(0, 5))
SeqFunc(Lambda(n, n**2), ([0, 5], 1))
```

flacjacket commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+        return self._eval_coeff(i)
+
+    def _eval_coeff(self, i):
+        raise NotImplementedError(filldedent(""" The _eval_coeff method should\
```
As a side note, you can do something like:
```
some_func("This is a long "
          "wrapping string that "
          "Python will put together correctly")
```
so you don't have to deal with dedenting long text

flacjacket commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+    def _dummy(self):
+        return ()
+
+    def coeff(self, i):
```
`i` is probably a bad name for this variable if it is supposed to be a value in the sequence, rather than an index

flacjacket commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+        -10
+        """
+        if self.start is S.NegativeInfinity:
+            initial = self.stop
```
You'll snag here if stop is positive infinity.

leosartaj replied:
``Range`` doesn't allow both end-points to be unbounded. I took a similar path. I will rethink on both end-points being unbounded.

leosartaj replied:
The reason I disallowed both end-points to be unbounded is due to ``__getitem__``. Now if a sequence is defined on ``(-oo, oo)``, what will 0th index be? I have tried to define 0th index as the starting point of the sequence. This is for convenience, say a sequence is defined in ``(5, 10)`` then 0th index will be value at point 5.

flacjacket replied:
Ah, yes, that makes sense.

flacjacket commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,522 @@
+
+
+class SeqPer(SeqExpr):
+    """Represents a periodical sequence
```
s/periodical/periodic

leosartaj committed:
Implemented _validate method in sequence classes

Other small changes:
* Got rid of _foo properties
(https://github.com/sympy/sympy/pull/9435#discussion_r31267208)
* Other minor changes (Typo and such)

leosartaj commented:
On the basis of suggestions by @jcrist @flacjacket 

Changes:
* Implemented a wrapper function that returns appropriate sequence.
* Improved validation of arguments (hopefully)
* Got rid of _foo properties
* Minor changes(typo and such)

leosartaj commented:
I have started working on operations now (starting with addition).

leosartaj committed:
Implemented SeqExprOp Class

Base class for operations on sequences

```
>>> from sympy.series.sequences import SeqExprOp, sequence
>>> from sympy.abc import n
>>> s1 = sequence(formula=(n**2, n), interval=(0, 10))
>>> s2 = sequence(periodical=(1, 2, 3), interval=(5, 10))
>>> s = SeqExprOp(s1, s2)
>>> s.gen
((n**2, n), (1, 2, 3))
>>> s.interval
[5, 10]
>>> s.length
6
```

leosartaj committed:
Implemented SeqAdd class

Represents addition of sequences
Rules:
    * The interval on which sequence is defined is the intersection
    of respective intervals of sequences.
    * Anything + EmptySequence, remains unchanged
    * Other rules are defined in _add methods of sequence classes

```
>>> from sympy import S, oo, SeqAdd, SeqPer, SeqFormula
>>> from sympy.abc import n
>>> SeqAdd(SeqPer((1, 2)), S.EmptySequence)
SeqPer((1, 2), ([0, oo), 1))
>>> SeqAdd(SeqPer((1, 2), (0, 5)), SeqPer((1, 2), (6, 10)))
EmptySequence()
>>> SeqAdd(SeqPer((1, 2)), SeqFormula(n**2))
SeqFormula((n**2, n), ([0, oo), 1)) + SeqPer((1, 2), ([0, oo), 1))
```

leosartaj commented:
I have implemented ``SeqAdd`` class. Addition is now supported. Though subtraction is still not. I have not yet implemented negation. Have a look @jcrist @flacjacket. Suggestions welcome.

leosartaj commented:
I have defined multiplication, subtraction. @jcrist @flacjacket

aktech commented on sympy/core/tests/test_args.py:
```diff
@@ -498,7 +497,7 @@ def test_sympy__sets__fancysets__Reals():
 def test_sympy__sets__fancysets__ImageSet():
     from sympy.sets.fancysets import ImageSet
-    from sympy import S, Lambda, Symbol
+    from sympy import S, Symbol
```
Perhaps symbols are already imported above, you can use that.

leosartaj replied:
I needed to import`` Lambda``, which was imported multiple times already. So, I took it out. Well, I will try to cleanup other imports as well. :+1:

aktech commented on sympy/core/tests/test_args.py:
```diff
@@ -2065,7 +2064,7 @@ def test_sympy__matrices__expressions__determinant__Determinant():
 def test_sympy__matrices__expressions__funcmatrix__FunctionMatrix():
     from sympy.matrices.expressions.funcmatrix import FunctionMatrix
-    from sympy import Lambda, symbols
+    from sympy import symbols
```
same here.

leosartaj committed:
Implemented SeqMul class, defined subtraction

Represents term-wise multiplication of sequences.
Handles multiplication of sequences only. For multiplication
with other objects see ``SeqBase.coeff_mul``.

Rules:
    * The interval on which sequence is defined is the intersection
    of respective intervals of sequences.
    * Anything * EmptySequence returns EmptySequence
    * Other rules are defined in _mul methods of sequence classes

```
>>> from sympy import S, oo, SeqMul, SeqPer, SeqFormula
>>> from sympy.abc import n
>>> SeqMul(SeqPer((1, 2)), S.EmptySequence)
EmptySequence()
>>> SeqMul(SeqPer((1, 2), (0, 5)), SeqPer((1, 2), (6, 10)))
EmptySequence()
>>> SeqMul(SeqPer((1, 2)), SeqFormula(n**2))
SeqPer((1, 2), ([0, oo), 1))*SeqFormula((n**2, n), ([0, oo), 1))
>>> SeqMul(SeqFormula(n**3), SeqFormula(n**2))
SeqFormula((n**5, n), ([0, oo), 1))
```

leosartaj commented:
I see ``sequences`` is a name already taken up and used in python a lot, in a different sense. I think we should change the name to maybe ``Streams``? Suggestion? @flacjacket @jcrist @toolforger

jcrist commented:
`Sequence` is fine. It has a different meaning in python, but it also has a specific mathematical meaning - which is what we're implementing here. `Streams` would be a bad choice IMO. I wouldn't worry about it.

leosartaj commented:
That's reassuring :+1:

leosartaj committed:
implemented pretty/latex printing

leosartaj commented:
I have implemented printing. This is my first interaction with the printing module. Let me know if I missed something or did the wrong way. @jcrist @flacjacket

leosartaj committed:
coeff method in SeqBase class is cached

leosartaj commented:
I think, it is now ready for review. @jcrist @flacjacket

jcrist commented:
I'm in the process of moving to Austin. Probably won't be able to look over this until Thursday.

flacjacket commented:
Great, I can take a look later today. Good luck on the move, Jim.

On Tue, Jun 2, 2015, 11:01 Sartaj Singh <notifications@github.com> wrote:

> I think, it is now ready for review. @jcrist <https://github.com/jcrist>
> @flacjacket <https://github.com/flacjacket>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9435#issuecomment-108000922>.
>

flacjacket commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1144 @@
+    is_SeqFunc = False
+    is_SeqFormula = False
+    is_SeqAdd = False
+    is_SeqMul = False
```
Is there a reason to define all of these like this rather than using `isinstance()`?

leosartaj replied:
There is no greater reason. I find it convenient that's all. Should I remove these or it's fine?

toolforger replied:
Am 04.06.2015 um 19:35 schrieb Sartaj Singh:
>> +    is_iterable = True
>> +
>> +    is_EmptySequence = False
>> +    is_SeqPer = False
>> +    is_SeqFunc = False
>> +    is_SeqFormula = False
>> +    is_SeqAdd = False
>> +    is_SeqMul = False
>
> There is no greater reason. I find it convenient that's all. Should I remove these or it's fine?

Please remove.
They're redundant, and redundancy means that any changes in the future 
will have to be applied consistently. Imagine somebody refactoring stuff 
and suddenly seeing breakage because they didn't know how the `is_*` 
attributes are supposed to work (subclasses have a universal and 
predefined way how they work, so people will know what to do with them).

Nonredundant `is_*` attributes will have to stay, of course.

leosartaj replied:
I don't think they are redundant in any way. Also, they are used abundantly in the SymPy's codebase (eg. sets module).

toolforger replied:
If the same information can be obtaining through `isinstance` checks, then yes they are redundant.
I know that this pattern is used abundantly, but that's not necessarily best practices; in fact it's considered an antipattern in the OO literature, and it got us into trouble already. (That doesn't mean that antipatterns are unavoidable; sometimes they're just the lesser of two evils. However, in the case of SymPy's `is_*` attributes, one of the deeper reasons we're having trouble with them is that we don't know which of them were unavoidable and which of them could be replaced with proper OO functions.)

Grand sweeping considerations aside, you need to document the semantics of each `is_*` property.
The documentation needs to be complete enough so that any Python programmer who sees a use of such an `is_*` property can instantly validate it by looking at the `is_*` property's docstring, *without looking at more than a handful of lines around the use of the `is_*` property*.
You'll need look at each use of any `is_*` property. Also, we'll have to review each such use and validate that it matches the docstring (not looking forward to that; essentially, such properties are convenient for coding but they turn into the opposite for review and maintenance).

leosartaj replied:
That looks like a valid point. I will remove them. Thanks!

flacjacket commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1144 @@
+from __future__ import print_function, division
+
+from sympy.core.expr import Expr
+from sympy.core.singleton import (S, Singleton)
```
Don't need the parentheses

flacjacket commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1144 @@
+    @property
+    def gen(self):
+        """Returns the generator for the sequence"""
+        raise NotImplementedError("(%s)._gen" % self)
```
Should be `"(%s).gen" % self`, similarly for other NotImplementedError's below

flacjacket commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1144 @@
+    sympy.series.sequences.SeqFunc
+    """
+    interval = kwargs.pop('interval', None)
+    if interval == None:
```
Checking against None should be `x is None` and `x is not None` rather than `x == None` and `not x == None`

flacjacket commented:
Other than my few comments, this looks fine

certik commented:
@leosartaj I played with the PR a bit, e.g.:
```
In [5]: SeqFormula(n**2, (-oo, 0))
Out[5]: […, 9, 4, 1, 0]

In [6]: SeqFormula(n**2, (0, oo))
Out[6]: [0, 1, 4, 9, …]

In [7]: SeqFormula(n**2, (0, oo))[:10]
Out[7]: [0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

In [8]: SeqFormula(n**2, (0, oo))[:50]
Out[8]: 
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 3
24, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 108
9, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 211
6, 2209, 2304, 2401]

In [9]: SeqFormula(n**2, (0, oo)) + SeqFormula(1/n, (0, oo))
Out[9]: [zoo, 2, 9/2, 28/3, …]

In [10]: SeqFormula(n**2, (0, oo)) + SeqFormula(1/n, (1, oo))
Out[10]: [2, 9/2, 28/3, 65/4, …]

In [11]: a = SeqFormula(n**2, (0, oo)) + SeqFormula(1/n, (1, oo))

In [12]: a
Out[12]: [2, 9/2, 28/3, 65/4, …]

In [13]: a.args
Out[13]: 
⎛⎛ 2   1   ⎞             ⎞
⎜⎜n  + ─, n⎟, ([1, ∞), 1)⎟
⎝⎝     n   ⎠             ⎠
```
and this is very good! Thanks @flacjacket and @jcrist for reviewing the PR, I don't have any further comments either, besides what you mentioned above.

@leosartaj how would one automatically obtain a formula for the Taylor series of e.g. a `sin(x)`?

leosartaj commented:
I did not completely understand you. ``Sequences`` are just used for representation they do not calculate anything.

```
>>> a = SeqFormula(n**2)
>>> a. formula
n**2
```
If this is what you mean.

leosartaj commented:
``Sequences`` will be used for representing series, like ``FormalPowerSeries``. I am still to implement series based classes.

certik commented:
So we currently have:
```
In [2]: sin(x).series(x, 0, 10)
Out[2]: 
     3     5     7       9           
    x     x     x       x       ⎛ 10⎞
x - ── + ─── - ──── + ────── + O⎝x  ⎠
    6    120   5040   362880         
```
You are right, I forgot, that your class doesn't represent the `x` yet, so let's evaluate at a given point, e.g.:
```
In [5]: [sin(x).series(x, 0, n).removeO().subs(x, 1) for n in range(1, 10)]
Out[5]: 
⎡                   101  101  4241  4241⎤
⎢0, 1, 1, 5/6, 5/6, ───, ───, ────, ────⎥
⎣                   120  120  5040  5040⎦
```
Essentially, each term is a partial sum of the type `1 - 1/6 + 1/120 - 1/5040 ...`. You can get the individual terms by:
```
In [8]: SeqFormula((-1)**n * 1 / factorial(1+2*n), (0, oo))
Out[8]: [1, -1/6, 1/120, -1/5040, …]
```
My question was whether there is an automated way to do that, so that I do not have to type the `(-1)**n * 1 / factorial(1+2*n)` part by hand.


A separate question is this: in limits, integrals etc. we specify the variable via the tupe `(n, 0, oo)`. However, this doesn't work for SeqFormula:
```
In [9]: SeqFormula((-1)**n * 1 / factorial(1+2*n), (n, 0, oo))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-af8ee1c1b6bd> in <module>()
----> 1 SeqFormula((-1)**n * 1 / factorial(1+2*n), (n, 0, oo))

/home/certik/repos/sympy/sympy/series/sequences.pyc in __new__(cls, gen, interval)
    436 
    437     def __new__(cls, gen, interval=(0, S.Infinity, 1)):
--> 438         bounds, step = _parse_interval(interval)
    439         if bounds is S.EmptySet:
    440             return S.EmptySequence

/home/certik/repos/sympy/sympy/series/sequences.pyc in _parse_interval(interval)
    400 
    401     if step in [S.NegativeInfinity, S.Infinity]:
--> 402         raise ValueError("step cannot be unbounded")
    403 
    404     return (Interval(start, stop), sympify(step))

ValueError: step cannot be unbounded
```
Why not?

leosartaj commented:
> My question was whether there is an automated way to do that, so that I do not have to type the (-1)**n * 1 / factorial(1+2*n) part by hand.

``SeqFormula`` needs explicit formula, there is no other way.

However, this issue can be addresed by a new ``TaylorSeries`` class. Now, there is no easy way to get a formula for a general function. On some initial thinking, maybe there can be a private function for functions, that returns the formula if it exists. Otherwise, it will resort to the original differentiation to calculate. Getting a formula and just plugging into ``SeqFormula`` will not be very optimized. I am still thinking on how to optimize so, it's easy to generate terms that depend on previous terms.

In the current form ``SeqFormula`` will be very suited to ``FormalPowerSeries`` though.

I will have to think more on ``TaylorSeries``. I am still deciding on the class structure. 

> A separate question is this: in limits, integrals etc. we specify the variable via the tupe (n, 0, oo). However, this doesn't work for SeqFormula:

I implemented ``SeqPer`` and ``SeqFunc`` that do not need a dummy variable. So, I thought this design to be more consistent. Interval remains same in all the types.

In my proposal I talked about ``SeriesX`` class. Now this class will use all the types of sequences, to allow users to generate different series representation. However, it does not calculate anything. Calculation will be handled by specific series classes.

leosartaj committed:
fixed typo's and other small changes

leosartaj commented:
I have addressed all the comments. Have a look @flacjacket

certik commented:
> I implemented `SeqPer` and `SeqFunc` that do not need a dummy variable. So, I thought this design to be more consistent. Interval remains same in all the types.

Indeed, `SeqPer` and `SeqFunc` do not have a symbol, so those are fine. But `SeqFormula` does, and you need to be able to tell it which symbol to sum over, don't you? I really think you should use the `(x, 0, 5)` tuple, like other such classes in sympy.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+from sympy.core.function import Lambda
+from sympy.core.add import Add
+from sympy.core.mul import Mul
+from sympy.core.compatibility import (range, integer_types, with_metaclass,\
```
No need to backslash inside parens.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+                start = 0
+            if stop is None:
+                stop = self.length
+            return [self.coeff(self._ith_point(i, index.step)) for i in\
```
No need to backslash inside parens.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+        elif len(interval) == 3:
+            start, stop, step = interval
+
+    if step == None:
```
All comparisons with `None` should use `is` or `is not`.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+    if step == None:
+        step = 1 # default
+
+    if start == None or stop == None:
```
Same here. I'm not going to tag all of them, please go through and cleanup these in all files. There's a few other places.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+            formula = (form1 + form2.subs(v2, v1), v1)
+            return SeqFormula(formula, self._intersect_interval(other))
+
+
```
extra blank line.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+        """See docstring of SeqBase.coeff_mul"""
+        coeff = sympify(coeff)
+        formula = self.formula * coeff
+        return SeqFormula((formula, self.variables[0]), \
```
No need for backslash inside parens. I'm not going to tag all of these, please go through and cleanup - there's a few other places this happens.

flacjacket replied:
As a note @leosartaj, you can check for stuff like this with a PEP 8 checker like flake8. Don't feel obligated to change everything it flags (especially long lines, error 501), but it will catch things like this, `is None`, etc.

leosartaj replied:
I will be careful the next time.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+
+    def _add(self, other):
+        """See docstring of SeqBase._add"""
+        if isinstance(other, SeqFunc)
```
Missing `:` after if.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+
+    def _mul(self, other):
+        """See docstring of SeqBase._mul"""
+        if isinstance(other, SeqFunc)
```
Missing `:` after if.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+                            " iterables of Sequences")
+        args = _flatten(args)
+
+        args = [a for a in args if not a is S.EmptySequence]
```
`a for a in args if a is not S.EmptySequence]`

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+                    # This returns None if s does not know how to add
+                    # with t. Returns the newly added sequence otherwise
+                    if new_seq is not None:
+                        new_args = [a for a in args if not a in (s, t)]
```
`[a for a in args if a not in (s, t)]`

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+                    # This returns None if s does not know how to multiply
+                    # with t. Returns the newly multiplied sequence otherwise
+                    if new_seq is not None:
+                        new_args = [a for a in args if not a in (s, t)]
```
`[a for a in args if a not in (s, t)]`

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+
+        return initial + i*step
+
+    def add(self, other):
```
Just a question of api design - do we need `add`, `sub` and `mul` methods defined like this? When would you use these instead of operators? Also, because `Sequence` looks like a collection, users may interpret `add` to add an element to the sequence, not do termwise addition. If you have a reason for these, then that's fine, otherwise I'd prefer to remove them in favor of just operators.

jcrist replied:
Also, why no `div`?

leosartaj replied:
Oh, that can get confusing for the users, I agree on that. I added them to provide somewhat similar interface to sets. They provide ``union``, ``intersection``, etc. I guess it makes sense there and not here. I will remove them.

leosartaj replied:
I couldn't really decide on how ``div`` should work. Plus I didn't find much of it's case. I say it will best to implement it incase we get a use for it.

leosartaj replied:
If you have some thoughts on ``div`` or find a need for it. Let me know.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+        while(i < self.length):
+            pt = self._ith_point(i)
+            yield self.coeff(pt)
+            i += 1
```
```
for i in range(self.length):
    pt = self._ith_point(i)
    yield self.coeff(pt)
```

leosartaj replied:
That' cleaner :+1:

leosartaj replied:
This will not work. As length can be oo. So it will cause issues. Just found out.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+                k = Dummy('k')
+                formula = Tuple(formula, k)
+            elif len(free) == 1:
+                formula = Tuple(formula, free.pop())
```
no need to pop, `free[0]` is clearer.

leosartaj replied:
``free`` is a ``set``.

jcrist replied:
ah, ok then.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+
+    Elements are generated by calling a function.
+    Only single argument functions are allowed.
+    Only SymPy's :class:`Lambda` functions are permitted.
```
Since only `Lambda` are allowed, and `Lambda` only works with sympy expressions, I feel that this whole class is redundant with `SeqFormula`. Thoughts?

leosartaj replied:
Yes, I agree. I mentioned it before, but implemented it nevertheless. I don't think it covers any extra use case. I will get rid of it.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+        """Generator for the sequence
+        returns a tuple of generators of all the argument sequences
+        """
+        return tuple([a.gen for a in self.args])
```
`tuple(a.gen for a in self.args)`. No need to create a list, and then put it in a `tuple`.

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+        """Sequence is defined on the intersection
+        of all the intervals of respective sequences
+        """
+        return Intersection(*[a.interval for a in self.args])
```
`Intersection(a.interval for a in self.args)`

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+        return Expr.__new__(cls, *args)
+
+    @staticmethod
+    def reduce(args):
```
Should this be public?

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+        val = 0
+        for a in self.args:
+            val += a.coeff(pt)
+        return val
```
`sum(a.coeff(pt) for a in self.args)`

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1123 @@
+        if len(args) == 0:
+            return S.EmptySequence
+
+        if Intersection(*[a.interval for a in args]) is S.EmptySet:
```
`Intersection(a.interval for a in args)`

jcrist commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1522,6 +1521,30 @@ def _print_Contains(self, e):
         else:
             return prettyForm(sstr(e))
 
+    def _print_SeqFormula(self, s):
```
`_print_SeqBase` could be used here instead - then you wouldn't need the additional methods for `SeqAdd`, etc...

leosartaj replied:
That will cause issues

```
In [1]: from sympy.series.sequences import SeqExpr

In [2]: s = SeqExpr(n**2)

In [3]: s
Out[3]: ---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-3-f4d5d0c0671b> in <module>()
----> 1 s

/home/leosartaj/projects/sympy/my/lib/python2.7/site-packages/IPython/core/displayhook.pyc in __call__(self, result)
    251             self.write_output_prompt()
    252             format_dict, md_dict = self.compute_format_data(result)
--> 253             self.write_format_data(format_dict, md_dict)
    254             self.update_user_ns(result)
    255             self.log_output(format_dict)

/home/leosartaj/projects/sympy/my/lib/python2.7/site-packages/IPython/core/displayhook.pyc in write_format_data(self, format_dict, md_dict)
    172         # newline, even if all the prompt separators are ''. This is the
    173         # standard IPython behavior.
--> 174         result_repr = format_dict['text/plain']
    175         if '\n' in result_repr:
    176             # So that multi-line strings line up with the left column of

KeyError: 'text/plain'
```

Well that's because it is abstract class, and doesn't have all the methods. So, I think I will need to define printers explicitly.

jcrist replied:
But no one should ever be creating instances of `SeqExpr` - that's just a base class. Not handling that case is fine, as it should never occur (correct me if I'm wrong).

leosartaj replied:
Well from user point of view it should never occur. But from a developer point of view, it will become a headache if everytime you initialize ``SeqExpr`` in isympy shell and you get a big error. I did have the case, when I had to initialize it, for argument checking. So, I think it's not a good idea.

jcrist replied:
The solution to this is not to use `isympy` for debugging the code then. The normal string printer works just fine.

leosartaj commented:
I will rethink on the api of ``SeqFormula`` @certik

leosartaj committed:
removed unnecessary backslashes, cleaned up functions

leosartaj committed:
got rid of mul, add and sub methods. Added docstring for __neg__

leosartaj commented:
I was thinking towards making ``Sequence`` api similar to ``Sum`` and other such classes.

Two options (Pseudo-code)

1. ``Sequence(generator, (n, 0, oo, 1))``
2. ``Sequence(generator, (n, 0, oo), 1)``

Let me know which you think is cleaner. I am confused here. @certik @flacjacket @jcrist

debugger22 commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1115 @@
+        args = _flatten(args)
+
+        # Multiplication of no sequences is EmptySequence
+        if len(args) == 0:
```
How about `if not args`?

leosartaj replied:
I think this will be confusing. args is a list. Now first impression of ``if not args`` is  something like a variable. I think len(args) serves a better purpose. Lot clearer to me.

jcrist replied:
The falsiness of lists and tuples is a common python idiom, and should be used to check for empty collections. It's also faster.

debugger22 commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,1115 @@
+        sequences if they can simplify themselves with any other constituent
+
+        Notes
+        =====
```
Perhaps a line here.

certik commented:
What is the purpose of the 1? I like the second option.

Sent from my mobile phone.
On Jun 5, 2015 12:25 PM, "Sartaj Singh" <notifications@github.com> wrote:

> I was thinking towards making Sequence api similar to Sum and other such
> classes.
>
> Two options (Pseudo-code)
>
>    1. Sequence(generator, (n, 0, oo, 1))
>    2. Sequence(generator, (n, 0, oo), 1)
>
> Let me know which you think is cleaner. I am confused here. @certik
> <https://github.com/certik> @flacjacket <https://github.com/flacjacket>
> @jcrist <https://github.com/jcrist>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9435#issuecomment-109387728>.
>

leosartaj commented:
1 defines the step size.
On 6 Jun 2015 10:12, "Ondřej Čertík" <notifications@github.com> wrote:
>
> What is the purpose of the 1? I like the second option.
>
> Sent from my mobile phone.
> On Jun 5, 2015 12:25 PM, "Sartaj Singh" <notifications@github.com> wrote:
>
> > I was thinking towards making Sequence api similar to Sum and other such
> > classes.
> >
> > Two options (Pseudo-code)
> >
> > 1. Sequence(generator, (n, 0, oo, 1))
> > 2. Sequence(generator, (n, 0, oo), 1)
> >
> > Let me know which you think is cleaner. I am confused here. @certik
> > <https://github.com/certik> @flacjacket <https://github.com/flacjacket>
> > @jcrist <https://github.com/jcrist>
>
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/sympy/sympy/pull/9435#issuecomment-109387728>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub.

certik commented:
I see. Yes, I really like the syntax  (x, a, b) for specifying the variable
and a range. It's consistent in SymPy and even in Mathematica.

Sent from my mobile phone.
On Jun 6, 2015 8:59 AM, "Sartaj Singh" <notifications@github.com> wrote:

> 1 defines the step size.
> On 6 Jun 2015 10:12, "Ondřej Čertík" <notifications@github.com> wrote:
> >
> > What is the purpose of the 1? I like the second option.
> >
> > Sent from my mobile phone.
> > On Jun 5, 2015 12:25 PM, "Sartaj Singh" <notifications@github.com>
> wrote:
> >
> > > I was thinking towards making Sequence api similar to Sum and other
> such
> > > classes.
> > >
> > > Two options (Pseudo-code)
> > >
> > > 1. Sequence(generator, (n, 0, oo, 1))
> > > 2. Sequence(generator, (n, 0, oo), 1)
> > >
> > > Let me know which you think is cleaner. I am confused here. @certik
> > > <https://github.com/certik> @flacjacket <https://github.com/flacjacket
> >
> > > @jcrist <https://github.com/jcrist>
> >
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <https://github.com/sympy/sympy/pull/9435#issuecomment-109387728>.
> >
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9435#issuecomment-109593534>.
>

jcrist commented:
I like option 1, with the step being optional. Then users like @certik can enter in a variable with a range like they do for other functionality (i.e. `(x, 0, 10)`), but can also add a step if needed. I like enclosing the whole variable, range, and step as one object, as they're related. Having the step be its own parameter seems awkward and unintuitive.

That said, I'm not sure why we need a step size - what was your use case for this @leosartaj?

leosartaj commented:
Well sequences are somewhat like ``Range``. ``Range`` does provide an option for step size. So, I also did.

```
>>> SeqFormula(n**2, (0, oo, 2))
[0, 4, 16, 36, ...]
```
Also, this type of sequences become easy to represent.

jcrist commented:
If you feel a step is a good option, I'm fine with that. But consider that your example above is equivalent to:

```
>>> SeqFormula((2*n)**2, (n, 0, oo))
[0, 4, 16, 36, ...]
```

certik commented:
If the step is optional, then I am fine with 1.

Sent from my mobile phone.
On Jun 6, 2015 9:16 AM, "Jim Crist" <notifications@github.com> wrote:

> I like option 1, with the step being optional. Then users like @certik
> <https://github.com/certik> can enter in a variable with a range like
> they do for other functionality (i.e. (x, 0, 10)), but can also add a
> step if needed. I like enclosing the whole variable, range, and step as one
> object, as they're related. Having the step be its own parameter seems
> awkward and unintuitive.
>
> That said, I'm not sure why we need a step size - what was your use case
> for this @leosartaj <https://github.com/leosartaj>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9435#issuecomment-109599207>.
>

leosartaj commented:
Yes, step is optional. So, I will go with 1.

jcrist commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1518,6 +1517,30 @@ def _print_Contains(self, e):
+            printset = (dots, s.coeff(stop - 3), s.coeff(stop - 2),
+                s.coeff(stop - 1), s.coeff(stop))
+        elif s.stop is S.Infinity or s.length > 4:
+            it = iter(s)
```
never used.

jcrist commented:
I just noticed you never define methods for string printing of your classes. This results in falling back on the default printer (which is fine in many cases). However, as a general rule, the string printed representation should be runnable code (ideally). This is not the case, due to you storing the args in a different format than they were input:

```python
>>> s = SeqFormula((n**2, n), (0, 5))
>>> s
SeqFormula((n**2, n), ([0, 5], 1))
>>> eval(repr(s))
<big error, trimmed traceback>
/home/jimmy/Code/sympy/sympy/core/sympify.py in sympify(a, locals, convert_xor, strict, rational, evaluate)
    270 
    271     if strict:
--> 272         raise SympifyError(a)
    273 
    274     if iterable(a):

SympifyError: SympifyError: [0, 5]
```

Solutions:
1. Add methods to the string printer
2. Store your args identically to their inputs

Either is fine, although the simpler the string printer the better. With the change in api we discussed above, this may work itself out.

toolforger commented:
TL;DR: I'm -1 on adding step size because the general design rule is to 
leave out any features that you aren't confident in being used, because 
the cost (nonzero) will outweigh the benefit (zero).

Full argument:

Generally speaking, making an API support more facilities always comes 
with wins and losses.
1) Win: Some use case that one wants to support.
2) Win: Even with no use case, APIs tend to find their application. So 
one can try it out and see whether people start using it. (You need to 
instrument the API with usage feedback, else you'll never know whether 
this win actually exists.)
3) Loss: More effort to write API-compatible code (i.e. subclasses in 
the case of an OO language).
4) Loss: More effort to check for potential incompatibilities with new 
extensions.
5) Loss: Higher risk of having to drop an extension because it cannot be 
made compatible.

In this context, we have:
1) Nope.
2) Dubious. An easy workaround exists, if the series becomes too 
cumbersome people will simply define a symbol and substitute.
3) Given that the Series stuff is new, it's likely that we'll want to 
write new subclasses. So we have a tangible loss here.
4) Hard to evaluate, since we have no idea how future extensions might 
look like. We should think more about potential extensions (and maybe 
use Series in practice), then reevaluate.
5) Essentially the same situation as with (4).

With nonexistent or dubious wins, I'd say that it's not worth even a 
potential loss, much less any losses that we don't know how to evaluate yet.

certik commented:
@leosartaj please ping me or @jcrist once you update the PR based on the feedback.

leosartaj commented:
Changes:

* Changed the api and made it similar to ``Sum``
* Removed step
* Removed ``SeqFunc``(redundant with ``SeqFormula``)

@jcrist @certik @flacjacket

leosartaj committed:
Removed SeqFunc, sequence api similar to Sum

jcrist commented:
Overall this looks very good! One new issue I noticed is that you're inheriting from `Expr`, not `Basic`. This allows the following to work, due to how python handles operators:

```
>>> n = symbols('n')
>>> seq = sequence(n**2)
>>> seq + n
< this errors, but with a poor error message >
>>> n + seq
n + SeqFormula(n**2, (n, 0, oo))
```

If we're not going to allow sequences to be part of expressions (I don't think we should), then sequences should be rooted in `Basic`, not `Expr`. Some arg checking in `SeqAdd` and `SeqMul` to detect non sequences would handle the poor error message issue.

Other than that, everything looks good!

jcrist commented on sympy/series/sequences.py:
```diff
@@ -0,0 +1,916 @@
+        """
+        return None
+
+    def coeff_mul(self, other):
```
what is this for? This should be supported by operators, not a special method.

leosartaj replied:
I tried.

```
>>> sequence(n**2) * 3
sequence(3*n**2) # works
```

I had trouble

```
>>> 3 *  sequence(n**2)
TyperError
```

Maybe ``__rmul__`` can be used here.

leosartaj replied:
Also, I am not very sure. It may cause some undesired results. Like multiplying ``sequence`` with a ``Tuple``.

jcrist replied:
If this only exists to make __neg__ work (all it's used with, for now), then you can just move this code inside __neg__, and remove the function. I really think operators should only work between two sequences, and that operations with sympy expressions should fail.

leosartaj replied:
I have add it's use in ``FourierSeries``.

leosartaj replied:
Yes only multiplying sequences with sequences should be allowed. I have kept this a separate method. It's works like ``map``. I am having trouble integrating it into ``__mul__``.

leosartaj replied:
Is it okay, if it's kept separate, and * knows how to multiply sequences with sequences only?

jcrist replied:
Ok, if it's needed elsewhere then it can stay as a method.  Perhaps later something prettier can be made.

leosartaj committed:
SeqBase inherits from Basic, improved error handling

leosartaj commented:
Changes:
* ``SeqBase`` now inherits from ``Basic``
* Only sequences can be added, subtracted and multiplied.

@jcrist

leosartaj commented:
Is this okay? @jcrist

jcrist commented:
Looks good, merging!

jcrist commented:
@leosartaj : can you please add something to the [release notes](https://github.com/sympy/sympy/wiki/Release-Notes-for-0.7.7) under major changes about the new sequences functionality. Thanks for you work!

leosartaj commented:
Sure, Thanks!
